File: frontend/src/api/address.ts
import { del, get, post, put } from './client'

export interface HomeAddress {
  id: number
  user_id: number
  alias: string
  address: string
  province?: string
  city?: string
  district?: string
  longitude: number
  latitude: number
  is_default: boolean
  note?: string
  created_at: string
  updated_at: string
}

export interface HomeAddressCreateInput {
  alias: string
  address: string
  province?: string
  city?: string
  district?: string
  longitude: number
  latitude: number
  is_default?: boolean
  note?: string
}

export type HomeAddressUpdateInput = Partial<HomeAddressCreateInput>

export const listAddresses = () =>
  get<{ list: HomeAddress[] }>('/addresses').then((r) => r.list)

export const createAddress = (body: HomeAddressCreateInput) =>
  post<HomeAddress>('/addresses', body)

export const updateAddress = (id: number, body: HomeAddressUpdateInput) =>
  put<HomeAddress>(`/addresses/${id}`, body)

export const deleteAddress = (id: number) => del<{ id: number }>(`/addresses/${id}`)

export const setDefaultAddress = (id: number) =>
  post<HomeAddress>(`/addresses/${id}/set-default`)

File: backend/internal/model/address.go
package model

import "time"

type HomeAddress struct {
	ID        int64     `json:"id"`
	UserID    int64     `json:"user_id"`
	Alias     string    `json:"alias"`
	Address   string    `json:"address"`
	Province  *string   `json:"province,omitempty"`
	City      *string   `json:"city,omitempty"`
	District  *string   `json:"district,omitempty"`
	Longitude float64   `json:"longitude"`
	Latitude  float64   `json:"latitude"`
	IsDefault bool      `json:"is_default"`
	Note      *string   `json:"note,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

type HomeAddressCreateInput struct {
	Alias     string   `json:"alias" binding:"required,max=64"`
	Address   string   `json:"address" binding:"required,max=512"`
	Province  *string  `json:"province" binding:"omitempty,max=32"`
	City      *string  `json:"city" binding:"omitempty,max=32"`
	District  *string  `json:"district" binding:"omitempty,max=32"`
	Longitude float64  `json:"longitude" binding:"required,gte=-180,lte=180"`
	Latitude  float64  `json:"latitude" binding:"required,gte=-90,lte=90"`
	IsDefault bool     `json:"is_default"`
	Note      *string  `json:"note" binding:"omitempty"`
}

type HomeAddressUpdateInput struct {
	Alias     *string  `json:"alias" binding:"omitempty,max=64"`
	Address   *string  `json:"address" binding:"omitempty,max=512"`
	Province  *string  `json:"province" binding:"omitempty,max=32"`
	City      *string  `json:"city" binding:"omitempty,max=32"`
	District  *string  `json:"district" binding:"omitempty,max=32"`
	Longitude *float64 `json:"longitude" binding:"omitempty,gte=-180,lte=180"`
	Latitude  *float64 `json:"latitude" binding:"omitempty,gte=-90,lte=90"`
	IsDefault *bool    `json:"is_default"`
	Note      *string  `json:"note"`
}
